Fix CPU custom operators for more than 32 qubits

stavros11 opened the pull request:
Based on the answer of [https://stackoverflow.com/questions/31465893/using-unsigned-long-long-as-iteration-range-in-for-loop-using-openmp](https://stackoverflow.com/questions/31465893/using-unsigned-long-long-as-iteration-range-in-for-loop-using-openmp), I think OpenMP doesn't like the `auto` option in parallelized for loops when this refers to an int64 integer.

Here I replace `auto` with `long` and this seems to fix the issue of not applying gates on CPU for more than 32 qubits.

stavros11 commented:
By the way, it would be useful to have some extreme case tests that catch issues like these which appear in large systems, however I am not sure if this can be integrated to the CI given the limited hardware resources of these machines. Perhaps we could run the "heavy" tests locally and skip them in CI?

scarrazza reviewed (commented):
@stavros11 thanks for this, I confirm that it fixes the issue.

However, given that we are using int64 everywhere, I think we should use int64 instead of long, for consistency. (in principle int64 is an alias of long and equivalent to size_t).

scarrazza commented:
> By the way, it would be useful to have some extreme case tests that catch issues like these which appear in large systems, however I am not sure if this can be integrated to the CI given the limited hardware resources of these machines. Perhaps we could run the "heavy" tests locally and skip them in CI?

Yes, sounds a good idea. In future we can setup a special runner on AWS.

stavros11 commented:
Thanks for the additional fixes. I replaced all `long` with `int64` as you suggested. I think all OMP loops are now fixed. Initial state uses `size_t` which I believe is also okay.
